container/controller.go

```diff
@@ -473,7 +474,51 @@ func (c *Controller) rpcHealthCheck() (chan struct{}, error) {
 			}
 		}
 	}()
+	return gone, nil
+}
+
+func isNFSMountStale(mountpoint string) bool {
+	if err := exec.Command("/bin/bash", "-c", fmt.Sprintf("read -t1 < <(stat -t '%s' 2>&-)", mountpoint)).Run(); err != nil {
+		if status, iscode := utils.GetExitStatus(err); iscode && status == 142 {
+			// EREMDEV; wait for NFS to come back
+			glog.Infof("Distributed storage temporarily unavailable. Waiting for it to return.")
+			return false
+
+		}
+		glog.Errorf("Mount point %s check had error (%s); considering stale", mountpoint, err)
+		return true
+	}
+	return false
+}
 
+// storageHealthCheck returns a channel that will close when the distributed
+// storage is no longer accessible (i.e., stale NFS mount)
+func (c *Controller) storageHealthCheck() (chan struct{}, error) {
+	gone := make(chan struct{})
+	mounts, err := mount.GetMounts()
+	if err != nil {
+		return nil, err
+	}
+	nfsMountPoints := []string{}
+	for _, minfo := range mounts {
+		if strings.HasPrefix(minfo.Fstype, "nfs") {
+			nfsMountPoints = append(nfsMountPoints, minfo.Mountpoint)
+		}
+	}
+	if len(nfsMountPoints) > 0 {
+		// Start polling
+		go func() {
+			for {
+				for _, mp := range nfsMountPoints {
```

Comment:
By which I mean: what would be a case where an nfs mount would be stale in
a container and we wouldn't care?
On Tue, Sep 1, 2015 at 12:24 PM Ian McCracken <ian.mccracken@gmail.com>
wrote:

> What would be an alternative?
> On Tue, Sep 1, 2015 at 12:22 PM Andrew Lapidas <notifications@github.com>
> wrote:
>
>> In container/controller.go
>> <https://github.com/control-center/serviced/pull/1925#discussion_r38446009>
>> :
>>
>> > +	gone := make(chan struct{})
>> > +	mounts, err := mount.GetMounts()
>> > +	if err != nil {
>> > +		return nil, err
>> > +	}
>> > +	nfsMountPoints := []string{}
>> > +	for _, minfo := range mounts {
>> > +		if strings.HasPrefix(minfo.Fstype, "nfs") {
>> > +			nfsMountPoints = append(nfsMountPoints, minfo.Mountpoint)
>> > +		}
>> > +	}
>> > +	if len(nfsMountPoints) > 0 {
>> > +		// Start polling
>> > +		go func() {
>> > +			for {
>> > +				for _, mp := range nfsMountPoints {
>>
>> Do we actually need to check every NFS mount point here?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/control-center/serviced/pull/1925/files#r38446009>.
>>
>